tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/op/TypedGradientAdapter.java

```diff
@@ -0,0 +1,88 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+=======================================================================
+
+*/
+package org.tensorflow.op;
+
+import static org.tensorflow.internal.c_api.global.tensorflow.StatusFromTF_Status;
+
+import java.lang.reflect.Constructor;
+import java.lang.reflect.InvocationTargetException;
+import java.util.List;
+import org.bytedeco.javacpp.PointerScope;
+import org.tensorflow.BaseGradientAdapter;
+import org.tensorflow.Graph;
+import org.tensorflow.Operand;
+import org.tensorflow.Output;
+import org.tensorflow.internal.c_api.NativeOperation;
+import org.tensorflow.internal.c_api.NativeOutputVector;
+import org.tensorflow.internal.c_api.NativeStatus;
+import org.tensorflow.internal.c_api.TF_Scope;
+import org.tensorflow.internal.c_api.TF_Status;
+
+/** A native adapter for {@link CustomGradient}. */
+final class TypedGradientAdapter<T extends RawOpInputs<?>> extends BaseGradientAdapter {
+
+  private final CustomGradient<T> gradient;
+  private final Class<T> opInputClass;
+  private final Constructor<T> ctor;
+
+  TypedGradientAdapter(CustomGradient<T> gradient, Class<T> opInputClass) {
+    super();
+    this.gradient = gradient;
+    this.opInputClass = opInputClass;
+    //noinspection unchecked
+    this.ctor = (Constructor<T>) this.opInputClass.getDeclaredConstructors()[0];
+  }
+
+  @Override
+  public NativeStatus call(
+      TF_Scope scope,
+      NativeOperation op,
+      NativeOutputVector grad_inputs,
+      NativeOutputVector grad_outputs) {
+    try (PointerScope pointerScope = new PointerScope()) {
+      Graph g = Graph.findGraphForPointer(scope.graph());
+      if (g == null) {
+        throw new IllegalStateException("No graph found for native gradient scope.");
+      }
+
+      Scope nativeScope = new GradientScope(scope, g);
+
+      Ops tf = new Ops(nativeScope);
+
+      List<Output<?>> gradInputs = BaseGradientAdapter.fromNativeOutputs(g, grad_inputs);
+
+      T rawOp = ctor.newInstance(BaseGradientAdapter.getGraphOp(g, op.node()));
+
+      BaseGradientAdapter.useDangerousLockedBuilders(g, true);
+      List<Operand<?>> gradOutputs = gradient.call(tf, rawOp, gradInputs);
+      BaseGradientAdapter.useDangerousLockedBuilders(g, false);
+
+      BaseGradientAdapter.putToNativeOutputs(gradOutputs, grad_outputs);
+
+    } catch (InvocationTargetException | InstantiationException | IllegalAccessException e) {
+      RuntimeException re =
+          new RuntimeException("Could not instantiate Op class " + opInputClass, e);
+      re.printStackTrace();
+      throw re;
+    } catch (Throwable t) {
```

Comment:
What other `Throwable` can come out of here?

Reply:
Nothing I can think of from the call, but the lambda could throw anything.  Leaking exceptions back to JNI didn't go well when I had it there (I forget exactly why, it may have been that the stack trace just wasn't printed).

Reply:
Given the block just rethrows the exception it's going back into JNI anyway. Shouldn't it catch the exception and construct a failing `TF_Status` then return that?

Reply:
If I remember right, I looked at that, but throwing the exception back to JNI forced a dump file to be generated, which is the only way you get a stacktrace that goes Java -> Native -> Java.  I'll check once I have native builds going.

Reply:
Throwing the exception seems to work fine through JNI, so I removed the print stacktrace.  However returning an error status causes the process to be killed in the native code, without printing a stacktrace, so I'm going to keep it as OK.

Reply:
Ok.